lib/providers/config_provider.dart

```diff
@@ -1,142 +1,370 @@
   }
+
+  void setWebSocketConnectionErrorPresent(bool errorPresent) {
+    _webSocketConnectionErrorPresent = errorPresent;
+    notifyListeners();
+  }
+
+  bool shouldNotify() {
+    if (!(showNotifications && (lastNotificationTimestamp == null || enoughTimeHavePassed()))) {
+      return false;
+    }
+    bool shouldNotify = false;
+    getAllWidgets().forEach((widget) {
+      if (quarantineWidgets.indexWhere((element) => element.id == widget.id) == -1) {
+        DashboardWidget notificationStateWidget =
+            _lastNotificationUpdateWidgetsState.firstWhere((element) => element.id == widget.id);
+        if (widget.content.containsKey(DashboardWidget.WIDGET_STATUS_KEY)) {
+          WidgetStatus widgetStatus =
+              EnumToString.fromString(WidgetStatus.values, widget.content[DashboardWidget.WIDGET_STATUS_KEY]);
+          WidgetStatus notificationWidgetStatus = EnumToString.fromString(
+              WidgetStatus.values, notificationStateWidget.content[DashboardWidget.WIDGET_STATUS_KEY]);
+          if (isErrorWidgetStatus(widgetStatus) ||
+              (!isErrorWidgetStatus(widgetStatus) && isErrorWidgetStatus(notificationWidgetStatus))) {
+            if (notificationStateWidget.content.containsKey(DashboardWidget.WIDGET_STATUS_KEY)) {
+              WidgetStatus notificationStateWidgetStatus = EnumToString.fromString(
+                  WidgetStatus.values, notificationStateWidget.content[DashboardWidget.WIDGET_STATUS_KEY]);
+              if (widgetStatus != notificationStateWidgetStatus) {
+                updateWidgetsInNotificationPayload(notificationStateWidget, widget);
+                shouldNotify = true;
+              }
+            } else {
+              shouldNotify = true;
+            }
+          }
+        }
+      }
+    });
+    if (shouldNotify) {
+      setNotificationPayload();
+    }
+    _lastNotificationUpdateWidgetsState = getAllWidgetsDeepCopy();
+    return shouldNotify;
+  }
+
+  bool enoughTimeHavePassed() =>
+      new DateTime.now().isAfter(lastNotificationTimestamp.add(new Duration(minutes: notificationFrequency)));
+
+  void updateWidgetsInNotificationPayload(DashboardWidget from, DashboardWidget to) {
+    WidgetStatusChange widgetStatusChange = new WidgetStatusChange(
+      from: DashboardWidget.deepCopy(from),
+      to: DashboardWidget.deepCopy(to),
+    );
+    int index = _widgetsInNotificationPayload.indexWhere((widgetStatusChange) => widgetStatusChange.from.id == from.id);
+    if (index == -1) {
+      _widgetsInNotificationPayload.add(widgetStatusChange);
+    } else {
+      _widgetsInNotificationPayload[index] = widgetStatusChange;
+    }
+  }
+
+  bool isErrorWidgetStatus(WidgetStatus widgetStatus) {
+    return widgetStatus == WidgetStatus.CHECKBOX_FAIL ||
+        widgetStatus == WidgetStatus.ERROR ||
+        widgetStatus == WidgetStatus.ERROR_CONFIGURATION ||
+        widgetStatus == WidgetStatus.ERROR_CONNECTION ||
+        widgetStatus == WidgetStatus.FAIL;
```

Comment:
could we have a constant SET of Error Statuses and just `Errors.contains(status)`?

Reply:
or even better: `WidgetStatus.isError(status)`

Reply:
WidgetStatus is an enum so to add method to enum I would need to use Dart 2.6 or higher and to do that I would need to update minimum sdk in yml, I don't know if this is a good choice:
https://stackoverflow.com/questions/38908285/add-methods-or-values-to-enum-in-dart
Moreover widgetStatus is tightly connected to widget - there would not be widget status without widget so I will do move isErrorWidget and  isWarningWidget methods to DashboardWidget so there will be widget.isError()

Reply:
will be changed in change-request-part1 PR

Reply:
[LATER] rename method to just `isError`